### PR TITLE
[TTS] Add modules for mel spectrogram codec

### PR DESCRIPTION
**Pull request by rlangman**
# What does this PR do ?

Add code and example for how to train audio codec while encoding mel spectrogram.

Note there is an extra "FullBandMelEncoder" definition which is not in the example config, but which I am adding to try to and keep inference working with some existing checkpoints.

**Collection**: [TTS]

# Changelog 
- Add codec modules for encoding mel spectrogram.
- Add example config file.

# Jenkins CI
To run Jenkins, a NeMo User with write access must comment `jenkins` on the PR.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

### CODE REVIEW

File: examples/tts/conf/audio_codec/speech_codec_44100.yaml [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+      _target_: nemo.collections.tts.modules.audio_codec_modules.MelSpectrogramProcessor
+      mel_dim: 80
+      sample_rate: ${sample_rate}
+      win_length: ${win_length}
+      hop_length: ${hop_length}
+
+  audio_decoder:
+    _target_: nemo.collections.tts.modules.audio_codec_modules.HiFiGANDecoder
+    up_sample_rates: ${up_sample_rates}
+    input_dim: 32
```
[the remark this answers is not in the document]
Comment (rlangman) [reply examples/tts/conf/audio_codec/speech_codec_44100.yaml]: Added.

File: examples/tts/conf/audio_codec/speech_codec_44100.yaml [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    base_channels: 1024 # This is double the base channels of HiFi-GAN V1, making it approximately 4x larger.
+
+  vector_quantizer:
+    _target_: nemo.collections.tts.modules.audio_codec_modules.GroupFiniteScalarQuantizer
+    num_groups: 8
```
[the remark this answers is not in the document]
Comment (rlangman) [reply examples/tts/conf/audio_codec/speech_codec_44100.yaml]: Added.

File: examples/tts/conf/audio_codec/speech_codec_44100.yaml [hunk 3]
```diff
@@ -0,0 +1,193 @@
+# This config contains the default values for training 24khz audio codec model
+# If you want to train model on other dataset, you can change config values according to your dataset.
+# Most dataset-specific arguments are in the head of the config file, see below.
+
+name: SpeechCodec
```
[the remark this answers is not in the document]
Comment (rlangman) [reply examples/tts/conf/audio_codec/speech_codec_44100.yaml]: Changed.

File: examples/tts/conf/audio_codec/speech_codec_44100.yaml [hunk 4]
```diff
@@ -0,0 +1,193 @@
+# This config contains the default values for training 24khz audio codec model
```
[the remark this answers is not in the document]
Comment (rlangman) [reply examples/tts/conf/audio_codec/speech_codec_44100.yaml]: Fixed.

File: nemo/collections/tts/modules/audio_codec_modules.py [hunk 5]
```diff
@@ -698,3 +715,415 @@ def decode(self, indices: torch.Tensor, input_len: torch.Tensor) -> torch.Tensor
...
+        mel_bands: List of mel spectrogram bands to encode.
+            Each list element is tuple of 2 elements with the start and end index of the mel features to use.
+        mel_processor: MelSpectrogramProcessor or equivalent class for computing the mel spectrogram from
+            input audio.
+        encoder_kwargs: Arguments for constructing encoder for each mel band.
+    """
+
+    def __init__(self, mel_bands: Iterable[Tuple[int, int]], mel_processor: NeuralModule, **encoder_kwargs):
+        super(MultiBandMelEncoder, self).__init__()
+        self.mel_bands = mel_bands
```
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/modules/audio_codec_modules.py:1096]: Added validation.

**Comment by rlangman** [comment]
jenkins

File: nemo/collections/tts/modules/audio_codec_modules.py [hunk 6]
```diff
@@ -698,3 +716,429 @@ def decode(self, indices: torch.Tensor, input_len: torch.Tensor) -> torch.Tensor
...
+
+
+class ResidualBlock(NeuralModule):
+    """
+    The residual block structure defined by the HiFi-GAN V1 and V2 configurations.
+    """
+
+    def __init__(
+        self,
+        in_channels: int,
```
Comment (nithinraok) [inline nemo/collections/tts/modules/audio_codec_modules.py:728]: minor: Are you following any published model here? Otherwise rename `in_channels` to `channels`? 

Doc string missing for args.

File: nemo/collections/tts/modules/audio_codec_modules.py [hunk 7]
```diff
@@ -698,3 +716,429 @@ def decode(self, indices: torch.Tensor, input_len: torch.Tensor) -> torch.Tensor
...
+        input_dim: Input dimension.
+        up_sample_rates: Rate to upsample for each decoder block. The product of the upsample rates will
+            determine the output frame rate. For example 8 * 8 * 2 * 2 = 256 samples per token.
+        base_channels: Number of filters in the first convolution. The number of channels will be cut in
+            half after each upsample layer.
+        in_kernel_size: Kernel size of the input convolution.
+        out_kernel_size: Kernel size of the output convolution.
+        resblock_kernel_sizes: List of kernel sizes to use in each residual block.
+        resblock_dilation_sizes: List of dilations to use in each residual block.
+        activation: Activation to use, defaults to leaky relu.
```
Comment (nithinraok) [inline nemo/collections/tts/modules/audio_codec_modules.py:865]: May be write: Activation in upsample layers? As Tanh is used for final activation.

File: nemo/collections/tts/modules/audio_codec_modules.py [hunk 8]
```diff
@@ -698,3 +716,429 @@ def decode(self, indices: torch.Tensor, input_len: torch.Tensor) -> torch.Tensor
...
+            n_window_stride=hop_length,
+            window_size=False,
+            window_stride=False,
+            n_fft=win_length,
+            mag_power=1.0,
+            log=True,
+            log_zero_guard_type="add",
+            log_zero_guard_value=log_guard,
+            mel_norm=None,
+            normalize=None,
```
Comment (nithinraok) [inline nemo/collections/tts/modules/audio_codec_modules.py:984]: curious: Are there any effects on the performance of encoder with normalization?

**Review by nithinraok — CHANGES_REQUESTED** [review]
Minor comments added. 

What about model class file? Would it be added later?

File: examples/tts/conf/audio_codec/mel_codec_44100.yaml [hunk 9]
```diff
@@ -0,0 +1,194 @@
+# This config contains the default values for training 44.1kHz audio codec model which encodes mel spectrogram
```
[the remark this answers is not in the document]
Comment (rlangman) [reply examples/tts/conf/audio_codec/mel_codec_44100.yaml:1]: Ideally both. I can add a second config file later if/when they are released.

File: examples/tts/conf/audio_codec/mel_codec_44100.yaml [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+    lr: 2e-4
+    betas: [0.8, 0.99]
+
+    sched:
+      name: ExponentialLR
+      gamma: 0.998
+
+trainer:
+  num_nodes: 1
+  devices: 1
```
[the remark this answers is not in the document]
Comment (rlangman) [reply examples/tts/conf/audio_codec/mel_codec_44100.yaml:166]: I would expect user to override these values manually if they are doing multi GPU/node training. Technically the config has only been tested on 8 GPUs..

I don't know of any training parameter that strictly needs to be changed when increasing the effective batch size. I didn't see any change when increasing learning rate with more devices, and have not tested different decay rates.

File: nemo/collections/tts/modules/audio_codec_modules.py (hunk 8 above)
Comment (rlangman) [reply nemo/collections/tts/modules/audio_codec_modules.py:984]: If you mean the `normalize` parameter which if I understand correctly does CMVN, then that would probably hurt compression performance because it would erase some information like channel effects.

If you mean `mel_norm`, then that goes to discussions long ago on whether we should use log(mel + 1.0), or the TTS standard definition of log(max(norm(mel), 1E-5)). If I remember correctly the results were mixed/inconclusive when we were using the original EnCodec losses, and after the changes in https://github.com/NVIDIA/NeMo/pull/7468 the version without normalization was generally better.

File: nemo/collections/tts/modules/audio_codec_modules.py [hunk 11]
```diff
@@ -60,6 +69,8 @@ def __init__(self, activation: str = "elu", channels: int = 1):
             self.activation = Snake(channels)
         elif activation == "elu":
             self.activation = nn.ELU()
+        elif activation == "lrelu":
+            self.activation = torch.nn.LeakyReLU()
```
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/modules/audio_codec_modules.py:75]: Added

File: nemo/collections/tts/modules/audio_codec_modules.py (hunk 6 above)
Comment (rlangman) [reply nemo/collections/tts/modules/audio_codec_modules.py]: Changed to `channels` and added doc string.

File: nemo/collections/tts/modules/audio_codec_modules.py (hunk 7 above)
Comment (rlangman) [reply nemo/collections/tts/modules/audio_codec_modules.py]: Changed.

**Comment by rlangman** [comment]
> What about model class file? Would it be added later?

This is trained using the same `audio_codec.py` model class.

**Comment by rlangman** [comment]
jenkins

**Review by nithinraok — APPROVED** [review]
LGTM, thanks
